<a href="https://colab.research.google.com/github/Danae-Munoz/Analisis_Data_Mineria-de-datos/blob/main/Evaluaci%C3%B3n_1_Danae_Mu%C3%B1oz.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Calificación y precio del vino**

**URL:** https://www.kaggle.com/datasets/budnyak/wine-rating-and-price

In [40]:
# Importación de las librerias necesarias
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb

In [33]:
# read data
red = pd.read_csv('Red.csv')
white = pd.read_csv('White.csv')
sparkling = pd.read_csv('Sparkling.csv')
rose = pd.read_csv('Rose.csv')

Esta dataset se divide en 4 csv y me gustaria trabajar con los datos como unico

In [34]:
red['WineStyle'] = 'red'
white['WineStyle'] = 'white'
sparkling['WineStyle'] = 'sparkling'
rose['WineStyle'] = 'rose'
df_wines =  pd.concat([red, white, sparkling, rose], ignore_index=True)

In [43]:
# Guardamos nuestro dataframe unificado
df_wines.to_csv('df_wines.csv', index=False)

In [45]:
# Para ver todas las columnas sin cortes
pd.set_option('display.max_columns', None)

# Para ver todas las filas (cuidado si el dataset es muy grande)
pd.set_option('display.max_rows', None)

In [76]:
# Ver informacion del dataframe
df_wines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13834 entries, 0 to 13833
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             13834 non-null  object 
 1   Country          13834 non-null  object 
 2   Region           13834 non-null  object 
 3   Winery           13834 non-null  object 
 4   Rating           13834 non-null  float64
 5   NumberOfRatings  13834 non-null  int64  
 6   Price            13834 non-null  float64
 7   Year             13834 non-null  float64
 8   WineStyle        13834 non-null  object 
dtypes: float64(3), int64(1), object(5)
memory usage: 972.8+ KB


In [78]:
# Visualizamos las primeras 5 filas
df_wines.head()

,Name,Country,Region,Winery,Rating,NumberOfRatings,Price,Year,WineStyle
0,Pomerol 2011,France,Pomerol,Château La Providence,4.2,100,95.00,2011.0,red
1,Lirac 2017,France,Lirac,Château Mont-Redon,4.3,100,15.50,2017.0,red
2,Erta e China Rosso di Toscana 2015,Italy,Toscana,Renzo Masi,3.9,100,7.45,2015.0,red
3,Bardolino 2019,Italy,Bardolino,Cavalchina,3.5,100,8.72,2019.0,red
4,Ried Scheibner Pinot Noir 2016,Austria,Carnuntum,Markowitsch,3.9,100,29.15,2016.0,red


In [77]:
# Visualizamos las ultimas 5 filas
df_wines.tail()

,Name,Country,Region,Winery,Rating,NumberOfRatings,Price,Year,WineStyle
13829,Côte des Roses Rosé 2019,France,Languedoc,Gérard Bertrand,3.9,1750,9.36,2019.0,rose
13830,M Rosé 2019,France,Côtes de Provence,Minuty,4.0,2046,12.85,2019.0,rose
13831,Whispering Angel Rosé 2019,France,Côtes de Provence,Château d'Esclans,4.2,2990,16.47,2019.0,rose
13832,Vinho Verde Rosé N.V.,Portugal,Vinho Verde,Casal Garcia,3.5,9493,4.35,2016.0,rose
13833,The Original Rosé N.V.,Portugal,Douro,Mateus,3.6,29531,5.99,2016.0,rose


In [60]:
# Devuelve la cantidad de filas y columnas respectivamente
df_wines.shape

(13834, 9)

Este dataframe tiene 13.834 filas y 9 columnas

In [69]:
# Observamos la columna y sus tipos de datos
df_wines.dtypes

,0
Name,object
Country,object
Region,object
Winery,object
Rating,float64
NumberOfRatings,int64
Price,float64
Year,float64
WineStyle,object


In [52]:
# Contamos la cantidad de nulos por columna
df_wines.isnull().sum()

,0
Name,0
Country,0
Region,0
Winery,0
Rating,0
NumberOfRatings,0
Price,0
Year,0
WineStyle,0


Podemos observar que el dataframe no tiene nulos, pero en el apartado de año, tiene N.V (sin año específico), vamos a solucionarlo

In [64]:
# Revisar qué valores raros tiene Year
df_wines['Year'].unique()

array([2011., 2017., 2015., 2019., 2016., 2014., 2013., 2018., 2012.,
       2010., 2007., 2005., 1999., 2002., 2008., 2006., 2009.,   nan,
       2000., 2004., 2001., 1996., 1998., 1990., 1995., 2003., 1993.,
       1988., 1989., 1997., 1992., 1991., 2020., 1961.])

In [63]:
# Transformar Year a numérico
df_wines['Year'] = pd.to_numeric(df_wines['Year'], errors='coerce')

In [99]:
# Ver cuántos nulos aparecieron
df_wines.isnull().sum()

,0
Country,0
Region,0
Rating,0
NumberOfRatings,0
Price,0
Year,0
WineStyle,0


Luego de la transformacion de la columna Year de categorica a numerica podemos observar que tenemos 744 nulos en total

**Desicion:** Al ser 744 nulos se decidio reemplazar los valores faltantes (NaN) de Year por un valor representativo, que en este caso seria la mediana. Se decidió no eliminar los registros debido a que implicaría la pérdida de información relevante.

In [72]:
# Ver la mediana de la columna Year
df_wines['Year'].median()

2016.0

In [73]:
# Imputar (rellenar los nulos)
df_wines['Year'] = df_wines['Year'].fillna(df_wines['Year'].median())

In [74]:
# Verificar que ya no existan nulos
df_wines.isnull().sum()

,0
Name,0
Country,0
Region,0
Winery,0
Rating,0
NumberOfRatings,0
Price,0
Year,0
WineStyle,0


Ahora que estamos seguros que no existen nulos pasemos a la siguiente fase

In [79]:
# Mostramos los nombres de las columnas
df_wines.columns

Index(['Name', 'Country', 'Region', 'Winery', 'Rating', 'NumberOfRatings',
       'Price', 'Year', 'WineStyle'],
      dtype='object')

**Realizaremos un mapeo de datos**
* Name: Categórico Nominal
* Country: Categórico Nominal
* Region: Categórico Nominal
* Winery: Categórico Nominal
* Rating: Numérico Continuo
* NumberOfRatings: Numérico Discreto
* Price: Numérico Continuo
* Year: Numérico Discreto
* WineStyle: Categórico Nominal



**Clasificación de Variables**

* **Categórico Nominal:** Categorías sin un orden o jerarquía específica.
  * Ejemplo: Country, Region, WineStyle, Winery.

* **Categórico Ordinal:** Categorías que siguen un orden lógico o escala.
  * Ejemplo: (En este dataset no se identifican variables ordinales explícitas)

* **Numérico Discreto:** Valores que surgen de un conteo y no admiten decimales.
  * Ejemplo: NumberOfRatings, Year.

* **Numérico Continuo:** Valores que resultan de una medición y pueden tener decimales infinitos.
  * Ejemplo: Price, Rating.

**Comenzamos el análisis de las columnas categoricas**

In [81]:
# Nos mostrará la cantidad de categoricas distintas unicas de Name
df_wines['Name'].value_counts()

,count
Name,
Chardonnay 2018,78
Sauvignon Blanc 2019,66
Sauvignon Blanc 2018,51
Cabernet Sauvignon 2017,43
Chardonnay 2017,41
Cabernet Sauvignon 2016,41
Cabernet Sauvignon 2015,35
Riesling Trocken 2018,27
Crianza 2016,27


In [82]:
# Ver cantidad de únicos de Name
df_wines['Name'].nunique()

10934

Podemos ver que la variable Name presenta una alta cardinalidad, con 10.934 valores únicos, lo que indica que la mayoría de los registros son distintos. Debido a esto, la variable no aporta valor predictivo, por lo que se estara evaluando su eliminacion del análisis

In [85]:
# Nos mostrará la cantidad de categoricas distintas unicas de Country
df_wines['Country'].value_counts()

,count
Country,
Italy,3919
France,3436
Spain,1533
Germany,1229
South Africa,846
United States,530
Austria,492
Chile,431
Portugal,336


In [86]:
# Ver cantidad de únicos de Country
df_wines['Country'].nunique()

33

La variable Country presenta un total de 33 categorías distintas. Esto indica que existe una diversidad moderada de países, lo que puede aportar valor al modelo al capturar diferencias geográficas en las características del vino.

In [87]:
# Nos mostrará la cantidad de categoricas distintas unicas de Region
df_wines['Region'].value_counts()

,count
Region,
Rioja,385
Stellenbosch,337
Pfalz,331
Toscana,307
Champagne,263
Südtirol - Alto Adige,227
Mosel,217
Saint-Émilion Grand Cru,207
Terre Siciliane,176


In [88]:
# Ver cantidad de únicos de Region
df_wines['Region'].nunique()

861

Podemos observar que la variable Region presenta 861 categorías distintas, lo que indica una alta cardinalidad. Esto puede generar problemas en el modelo al aumentar considerablemente la dimensionalidad del dataset. Por lo tanto, se estara evaluando su eliminacion

In [89]:
# Nos mostrará la cantidad de categoricas distintas unicas de Winery
df_wines['Winery'].value_counts()

,count
Winery,
Markus Molitor,73
Errazuriz,58
Torres,54
Joseph Drouhin,51
Gaja,42
M. Chapoutier,42
E. Guigal,36
Antinori,36
Paul Jaboulet Aîné,36


In [90]:
# Ver cantidad de únicos de Winery
df_wines['Winery'].nunique()

3505

Podemos observar que la variable Winery presenta un total de 3.505 categorías distintas, lo que indica una alta cardinalidad. Esto implica que la mayoría de las bodegas son únicas. Debido a esto, la variable no aporta valor predictivo relevante en relación con su complejidad, por lo que se evaluara eliminarla del análisis.

In [91]:
# Nos mostrará la cantidad de categoricas distintas unicas de WineStyle
df_wines['WineStyle'].value_counts()

,count
WineStyle,
red,8666
white,3764
sparkling,1007
rose,397


In [92]:
# Ver cantidad de únicos de WineStyle
df_wines['WineStyle'].nunique()

4

Podemos observar que la variable WineStyle presenta un total de 4 categorías distintas: red, white, sparkling y rose. Esto indica que la variable tiene una baja cardinalidad, lo que la hace adecuada para ser utilizada en el modelo.

In [97]:
df_wines = df_wines.drop(columns=['Name', 'Winery'])

**Se decidió eliminar las columnas Name y Winery debido a su alta cardinalidad, ya que presentan una gran cantidad de categorías únicas, lo que podría aumentar innecesariamente la dimensionalidad del dataset y afectar negativamente el rendimiento del modelo.**

In [101]:
# Verificar que se eliminaron
df_wines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13834 entries, 0 to 13833
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country          13834 non-null  object 
 1   Region           13834 non-null  object 
 2   Rating           13834 non-null  float64
 3   NumberOfRatings  13834 non-null  int64  
 4   Price            13834 non-null  float64
 5   Year             13834 non-null  float64
 6   WineStyle        13834 non-null  object 
dtypes: float64(3), int64(1), object(3)
memory usage: 756.7+ KB


In [ ]:
#------------------------------------------------------------

**Estadistica descriptiva a las variables numericas**

In [98]:
df_wines.describe()

,Rating,NumberOfRatings,Price,Year
count,13834.000000,13834.000000,13834.000000,13834.000000
mean,3.865664,428.322466,33.024850,2015.556961
std,0.296427,1838.413812,70.899893,3.188885
min,2.200000,25.000000,3.150000,1961.000000
25%,3.700000,56.000000,9.902500,2015.000000
50%,3.900000,129.000000,15.950000,2016.000000
75%,4.100000,336.000000,32.500000,2018.000000
max,4.900000,94287.000000,3410.790000,2020.000000


Se identifican 4 variables numericas lo cual es correcto con respecto a lo que identificamos antes

**Importaremos las librerias para realizar el encoder a Country, Region y WineStyle**

In [95]:
pip install category_encoders

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.9/85.9 kB 4.0 MB/s eta 0:00:00


In [102]:
from sklearn.preprocessing import OneHotEncoder
import category_encoders as ce

In [106]:
# Aplicar OneHotEncoder a Country y WineStyle
# Esto convierte esas columnas categóricas en columnas numéricas.
ohe = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

ohe_encoded = ohe.fit_transform(df_wines[['Country', 'WineStyle']])

Esto convierte las columnas categóricas Country y WineStyle en columnas numéricas.

In [107]:
# Convertir este resultado a DataFrame
ohe_df = pd.DataFrame(
    ohe_encoded,
    columns=ohe.get_feature_names_out(['Country', 'WineStyle']),
    index=df_wines.index
)

Esto transforma el resultado del encoder en una tabla normal con nombres de columnas claros.

In [108]:
# Aplicar Binary Encoding a Region
be = ce.BinaryEncoder(cols=['Region'])

region_encoded = be.fit_transform(df_wines[['Region']])

Convierte Region en varias columnas numéricas, pero sin generar 861 columnas como haría OneHot.

In [109]:
# Unir todo en un nuevo dataframe final
df_final = pd.concat(
    [
        df_wines.drop(columns=['Country', 'WineStyle', 'Region']),
        ohe_df,
        region_encoded
    ],
    axis=1
)

Para poder utilizar las variables categóricas en el modelo, fue necesario transformarlas a formato numérico mediante técnicas de encoding.

En el caso de las variables Country y WineStyle, se aplicó One-Hot Encoding, ya que cuentan con un número reducido de categorías. Este método consiste en crear una columna por cada categoría, permitiendo representar cada valor de forma independiente sin generar conflictos en el modelo.

Por otro lado, la variable Region presenta una gran cantidad de categorías (alta cardinalidad). Si se aplicara One-Hot Encoding, se generarían demasiadas columnas, lo que afectaría el rendimiento del modelo. Por esta razón, se utilizó Binary Encoding, una técnica que permite representar las categorías utilizando un menor número de columnas mediante combinaciones binarias, logrando así reducir la dimensionalidad del dataset.

In [110]:
df_final.head()

,Rating,NumberOfRatings,Price,Year,Country_Argentina,Country_Australia,Country_Austria,Country_Brazil,Country_Bulgaria,Country_Canada,Country_Chile,Country_China,Country_Croatia,Country_Czech Republic,Country_France,Country_Georgia,Country_Germany,Country_Greece,Country_Hungary,Country_Israel,Country_Italy,Country_Lebanon,Country_Luxembourg,Country_Mexico,Country_Moldova,Country_New Zealand,Country_Portugal,Country_Romania,Country_Slovakia,Country_Slovenia,Country_South Africa,Country_Spain,Country_Switzerland,Country_Turkey,Country_United Kingdom,Country_United States,Country_Uruguay,WineStyle_red,WineStyle_rose,WineStyle_sparkling,WineStyle_white,Region_0,Region_1,Region_2,Region_3,Region_4,Region_5,Region_6,Region_7,Region_8,Region_9
0,4.2,100,95.00,2011.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0,0,0,0,0,0,0,0,0,1
1,4.3,100,15.50,2017.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0,0,0,0,0,0,0,0,1,0
2,3.9,100,7.45,2015.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0,0,0,0,0,0,0,0,1,1
3,3.5,100,8.72,2019.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0,0,0,0,0,0,0,1,0,0
4,3.9,100,29.15,2016.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0,0,0,0,0,0,0,1,0,1


In [112]:
print(f"Total de columnas del nuevo dataframe: {len(df_final.columns)}")

Total de columnas del nuevo dataframe: 51
